### ISSUE TITLE
`string tabulate` command to replicate `column` util but take into account ANSI escapes (i.e. string length -V)

### ISSUE THREAD

**Issue by vergenzt** [fish-shell/fish-shell#9396]
**TL;DR:** It'd be nice if there was a `string tabulate` command to replicate [`column -t`](https://man.archlinux.org/man/column.1.en) but take into account the _visible_ length of each table cell (i.e. via `string length --visible`). I think it'd be a great fit for fish shell's string library.

Refs:
- [text processing - A shell tool to “tablify” input data containing ANSI escape codes - Unix & Linux Stack Exchange](https://unix.stackexchange.com/questions/121107/a-shell-tool-to-tablify-input-data-containing-ansi-escape-codes)
- [text processing - Why the output of "column" is misaligned with a ANSI colored input? - Unix & Linux Stack Exchange](https://unix.stackexchange.com/questions/340091/why-the-output-of-column-is-misaligned-with-a-ansi-colored-input)

Use case this came up in: I'm trying to write a fish helper function for my dotfiles to "summarize" the git states of all repositories in a directory. Naturally, colors are useful here. I'm using [starship](https://starship.rs/)'s [git](https://starship.rs/config/#git-branch) modules to produce the output since those are already tailored towards summarization & information density. It'd be nice to not have to strip the colors out in order to align the components with each other and make a "directories" (each row) x "starship modules" (each column) table!

## Example of what I mean

Code that exists now:
```console
$ set table_data this_foo_string_is_longer_than_8_characters bar baz booz

$ set -S table_data
$table_data: set in global scope, unexported, with 4 elements
$table_data[1]: |this_foo_string_is_longer_than_8_characters|
$table_data[2]: |bar|
$table_data[3]: |baz|
$table_data[4]: |booz|

$ paste (string length $table_data | psub) (string length --visible $table_data | psub)
43      43
3       3
3       3
4       4

$ printf %s\t%s\n $table_data | column -t
this_foo_string_is_longer_than_8_characters  bar
baz                                          booz

$ set table_data_color (set_color yellow)$table_data[1](set_color normal) (set_color blue)$table_data[2](set_color normal) $table_data[3..4]

$ set -S table_data_color
$table_data_color: set in global scope, unexported, with 4 elements
$table_data_color[1]: |\e[33mthis_foo_string_is_longer_than_8_characters\e(B\e[m|
$table_data_color[2]: |\e[34mbar\e(B\e[m|
$table_data_color[3]: |baz|
$table_data_color[4]: |booz|

$ paste (string length $table_data_color | psub) (string length --visible $table_data_color | psub)
54      43
14      3
3       3
4       4

$ printf %s\t%s\n $table_data_color | column -t
this_foo_string_is_longer_than_8_characters  bar
baz                                                  booz
```

What I'd like to exist:
```console
$ printf %s\t%s\n $table_data_color | string tabulate
this_foo_string_is_longer_than_8_characters  bar
baz                                          booz
```

---

I'm working on writing a helper function to do this myself, but figured I'd create an issue for the problem first. 🙂 Is this something the team would be open to merging?

**Comment by vergenzt** [comment]
Oh my goodness and I just noticed that `string pad` already does even more of the work! "The output is padded to the maximum width of all input strings." So really the _only_ thing missing from this is applying `string pad` to _each column_ separated by a given delimiter, then re-concating the columns.

**Comment by vergenzt** [comment]
(Just linking to #206 here cause that tripped me up while trying to implement this)

**Comment by vergenzt** [comment]
I wrote a pure fish function version of this. 🙂 Keeping it in my dotfiles for now... Lmk if there's interest in getting this merged. (Not sure how that would work since the other `string` utils are written in C, aren't they?)

```fish
function string-tabulate
  set _flag_delimiter \t
  set _flag_padchar ' '
  set _flag_out_delimiter ' '

  set _args 'd/delimiter=' 'c/padchar=' 'D/out-delimiter=' 'n/limit=!_validate_int --min=1' 
  argparse $args 'print-arg-specs' -- $argv

  if set -q _flag_print_arg_specs
    string collect $_args
    return
  end

  if not count $argv >/dev/null
    while read --line next_argv
      set -a argv $next_argv
    end
  end

  set n_rows (count $argv); or return
  set n_cols # initially empty. (we compute the max n_cols incrementally.)
  set col_vars # initially empty. (we expand this to contain the variable name for each column incrementally.)

  for i in (seq $n_rows)
    set row_str $argv[$i]
    set row (string split -- $_flag_delimiter $row_str)
    set n_cols_in_row (count $row)

    # expand $col_vars if needed
    for j in (seq $n_cols $n_cols_in_row)
      set col_vars[$j] col_$j
    end
    for j in (seq $n_cols_in_row)
      set $col_vars[$j][$i] $row[$j]
    end

    set n_cols (math max '(' (string join -- , $n_cols $n_cols_in_row ) ')')
  end

  # pad each column except the last
  set last_col_to_pad (
    if set -q _flag_limit
      math "min($n_cols, $_flag_limit + 1) - 1"
    else
      math "$n_cols - 1"
    end
  )
  for j in (seq $last_col_to_pad)
    set $col_vars[$j] (string pad --right --char=$_flag_padchar -- $$col_vars[$j])
  end

  # print each row
  for i in (seq $n_rows)
    for j in (seq $n_cols)
      printf %s $$col_vars[$j][$i]
      if test $j -eq $n_cols
        echo
      else
        printf %s $_flag_out_delimiter
      end
    end
  end
end

```

**Comment by Nahor** [comment]
Is this still an issue? `column -t` works for me:

<img width="449" height="56" alt="Image" src="https://github.com/user-attachments/assets/02c2c0be-e61e-4e44-b1c6-3193ff086b1b" />

**Comment by vergenzt** [comment]
> Is this still an issue? `column -t` works for me:
> 
> <img width="449" height="56" alt="Image" src="https://github.com/user-attachments/assets/02c2c0be-e61e-4e44-b1c6-3193ff086b1b" />
> 

@Nahor What version of `column` is that? I'd wager it's the GNU implementation, which is often much more feature-ful than other implementations. E.g. MacOS ships with the BSD versions of these tools  which often have a much more limited feature set.

Good to know that some implementation(s) do calculate a visible width though! Thanks for sharing.

**Comment by Nahor** [comment]
It's from `util-linux`, which I assume is what you mean by the "GNU version".

It should also be the most used one. My thinking though is that formatting column document is fairly niche to begin with, few people need/use it. And with the most common version of `column` handling invisible data correctly, the need of `string tabulate` would be extremely small. So I don't think this is worth adding to fish, and it were, it could open the flood gate for duplicating a lot of other existing less niche tools/functionality.

**Comment by vergenzt** [comment]
1. @Nahor the version of `column` which ships with MacOS does _not_ handle ANSI escapes (as of MacOS 26.4.1 -- `man column` on my machine lists a date of 2004-07-29 🧓) -- which was why I created this helper for myself in the first place.

   <img width="1007" height="131" alt="Image" src="https://github.com/user-attachments/assets/8fd81c71-4b23-46ab-b71d-06295e2be87a" />

   That's going to be a pretty sizable chunk of developers who do not have ready access to a version of `column` which handles ANSI escapes. (I didn't even know that such a version existed!)



2. > _I don't think this is worth adding to fish, and it were, it could open the flood gate for duplicating a lot of other existing less niche tools/functionality._

   "Duplicating" (i.e. standardizing and simplifying) other existing tools/functionality into fish-specific forms seems to be the _whole purpose of the `string` util_. The _majority_ its existing subcommands already duplicate other tools!

   E.g.
   * `string join` is equivalent to `paste`
   * `string join0` is equivalent to `tr '\n' '\0'`
   * `string length` is equivalent to `wc -c`
   * `string lower` is equivalent to `tr '[[:upper:]]' '[[:lower:]]'`
   * `string match` can almost always be substituted for `grep` and/or `sed`
   * `string pad` can be largely replicated with `printf`
   * `string repeat` can be mimicked with `yes`/`head`/`tr`
   * `string replace` is equivalent to `sed` (and arguably less powerful!)

   ... and so on.

   **To me it seems like a `string tabulate` subcommand which does one thing and does it right by default** (what does `column` do without the `-t` flag?!)... **would fit right in. 🙂**

**Comment by Nahor** [comment]
> pretty sizable chunk of developers who do not have ready access to a version of column

Define "sizable chunk"? If ANSI escape sequences was that necessary, then your version would be handling it already.
As for "ready access", if one installed fish, it's likely one also has `macports` or `homebrew` installed, which means it's only a small step to install `util-linux` as well.

> The majority its existing subcommands already duplicate other tools!

The issue is not the duplication in and of itself. The issue is the "why" it is duplicated. The existing `string` sub-commands exist because they are very basic functionalities that most scripts need to use at some point or another, to process inputs/variables/arguments, and often enough in loops, i.e. repeatedly. Having to call external commands each time adds some noticeable amount of time just to spawn those processes over and over, and then to pipe the data around and parse it. It's the same reason some other builtins exist as well (`echo`, `printf`, `true`, `false`, `test`, ...)
`column` is typically a one-time call inside a script, so the cost of that one spawn is unnoticeable. Even more so for ANSI escape sequence handling since that means the output is targeted at slow humans, not fast computers.

The fact that `column` is useful or not is beside the point in my opinion. `curl`/`wget` are useful too, much more so than `column`. So is a basic text editor such as `nano` or `vi`. Yet fish does not have those either.

> does it right by default (what does column do without the -t flag?!)

Since `column` is about formatting data, and worse, formatting for human consumption, "do it right"/"default" is inherently subjective (and the fact that you don't know what `column` does without `-t` only emphasizes that). I mean, just look at all the formatting options that `column` has.
In contrast, the `string` subcommands are all fairly simple, with few options, and cover 99% of the use cases.

> would fit right in. 

Maybe, but as they say "just because you can doesn't mean you should". I still don't think it's useful enough, to enough people, to be worth adding to fish. I still think it is better served by being its own dedicated utility.
